In [22]:
import pandas as pd
import numpy as np
import seaborn as sns

1. Attendance Data

Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

name

Billy    0.5250

Jane     0.6875

John     0.9125

Sally    0.7625

Name: grade, dtype: float64

In [23]:
attendance = pd.read_csv('untidy-data/attendance.csv')
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [45]:
# combine all columns except 'Unnamed: 0' into 'date' and hold result in 'grade'
att_melt = attendance.melt(id_vars=['Unnamed: 0'], var_name='date', value_name='grade')


In [34]:
att_melt.head()

Unnamed: 0        date grade
0      Sally  2018-01-01     P
1       Jane  2018-01-01     A
2      Billy  2018-01-01     A
3       John  2018-01-01     P
4      Sally  2018-01-02     T

In [44]:
# rename columns
att_melt.columns = ['student', 'date', 'grade']
att_melt.head()

student        date  grade
0   Sally  2018-01-01    1.0
1    Jane  2018-01-01    0.0
2   Billy  2018-01-01    0.0
3    John  2018-01-01    1.0
4   Sally  2018-01-02    0.9

In [42]:
# rename grade values to numbers
att_melt= att_melt.replace(['P', 'T', 'H', 'A'], [1, .9, .5, 0])

In [37]:
att_melt.head()

student        date  grade
0   Sally  2018-01-01    1.0
1    Jane  2018-01-01    0.0
2   Billy  2018-01-01    0.0
3    John  2018-01-01    1.0
4   Sally  2018-01-02    0.9

In [40]:
# attendance percentage for each student
att_melt.groupby('student').grade.mean()

student
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64

2. Coffee Levels

a. Read the coffee_levels.csv file.

In [50]:
coffee = pd.read_csv('untidy-data/coffee_levels.csv')
coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

b. Transform the data so that each carafe is in it's own column.

In [51]:
# pivotl table
coffee_pivot = coffee.pivot_table(index = ['hour'],
                    columns = 'coffee_carafe',
                    values = 'coffee_amount').reset_index()

In [54]:
coffee_pivot.head()

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947

c. Is this the best shape for the data?


In [55]:
# yes

3. Cake Recipes 

a. Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [56]:
cake = pd.read_csv('untidy-data/cake_recipes.csv')
cake.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

b. Tidy the data as necessary.


In [63]:
# multiple values in the same column so split the recipe:position column 
# and add values to two new columns

In [60]:
cake[['recipe', 'position']] = cake['recipe:position'].str.split(':', expand = True)

In [61]:
cake.head()

recipe:position        225        250        275        300 recipe position
0        a:bottom  61.738655  53.912627  74.414730  98.786784      a   bottom
1           a:top  51.709751  52.009735  68.576858  50.228470      a      top
2        b:bottom  57.095320  61.904369  61.196980  99.248541      b   bottom
3           b:top  82.455004  95.224151  98.594881  58.169349      b      top
4        c:bottom  96.470207  52.001358  92.893227  65.473084      c   bottom

In [64]:
# drop the the split column

In [65]:
cake.drop(columns = 'recipe:position', inplace = True)

In [66]:
cake.head()

225        250        275        300 recipe position
0  61.738655  53.912627  74.414730  98.786784      a   bottom
1  51.709751  52.009735  68.576858  50.228470      a      top
2  57.095320  61.904369  61.196980  99.248541      b   bottom
3  82.455004  95.224151  98.594881  58.169349      b      top
4  96.470207  52.001358  92.893227  65.473084      c   bottom

In [67]:
# melt the data

In [72]:
cake_melt = cake.melt(id_vars = ['recipe', 'position'], var_name = 'temperature', value_name = 'tastey_score')
cake_melt.head()

recipe position temperature  tastey_score
0      a   bottom         225     61.738655
1      a      top         225     51.709751
2      b   bottom         225     57.095320
3      b      top         225     82.455004
4      c   bottom         225     96.470207

c. Which recipe, on average, is the best? 

Answer: recipe b

In [78]:
# group by the recipe then find the average tastiness score
cake_melt.groupby(['recipe']).tastey_score.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: tastey_score, dtype: float64

d. Which oven temperature, on average, produces the best results? 275


Answer: 275

In [80]:
# Group by temperature then find the average tastiness score
cake_melt.groupby(['temperature']).tastey_score.mean()

temperature
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: tastey_score, dtype: float64

e. Which combination of recipe, rack position, and temperature gives the best result? 

Answer: recipe b, bottom rack, 300 degrees

In [83]:
cake_melt.groupby(['temperature', 'recipe', 'position']).tastey_score.mean()

temperature  recipe  position
225          a       bottom      61.738655
                     top         51.709751
             b       bottom      57.095320
                     top         82.455004
             c       bottom      96.470207
                     top         71.306308
             d       bottom      52.799753
                     top         96.873178
250          a       bottom      53.912627
                     top         52.009735
             b       bottom      61.904369
                     top         95.224151
             c       bottom      52.001358
                     top         82.795477
             d       bottom      58.670419
                     top         76.101363
275          a       bottom      74.414730
                     top         68.576858
             b       bottom      61.196980
                     top         98.594881
             c       bottom      92.893227
                     top         92.098049
             d       bot

In [ ]:
# 300       b       bottom      99.248541